h2o-persist-hdfs/src/main/java/water/persist/security/HdfsDelegationTokenRefresher.java

```diff
@@ -0,0 +1,215 @@
+        _executor.scheduleAtFixedRate(this, 0, interval, TimeUnit.MILLISECONDS);
+    }
+
+    private static void log(String s, Exception e) {
+        System.out.println("TOKEN REFRESH: " + s);
+        if (e != null) {
+            e.printStackTrace(System.out);
+        }
+    }
+
+    @Override
+    public void run() {
+        if (Paxos._cloudLocked && !(H2O.CLOUD.leader() == H2O.SELF)) {
+            // cloud is formed the leader will take of subsequent refreshes
+            _executor.shutdown();
+            return;
+        }
+        try {
+            Credentials creds = refreshTokens(loginAuthUser());
+            distribute(creds);
+        } catch (IOException | InterruptedException e) {
+            log("Failed to refresh token.", e);
+        }
+    }
+
+    private Credentials refreshTokens(UserGroupInformation tokenUser) throws IOException, InterruptedException {
+        return tokenUser.doAs((PrivilegedExceptionAction<Credentials>) () -> {
+            Credentials creds = new Credentials();
+            try {
+                Token<?>[] tokens = fetchDelegationTokens(_authUser, creds);
+                log("Fetched delegation tokens: " + Arrays.toString(tokens), null);
+            } catch (IOException e) {
+                log("Failed to fetch delegation tokens", e);
+            }
+            return creds;
+        });
+    }
+
+    private UserGroupInformation loginAuthUser() throws IOException, InterruptedException{
+        log("Log in from keytab as " + _authPrincipal, null);
+        UserGroupInformation realUser = UserGroupInformation.loginUserFromKeytabAndReturnUGI(_authPrincipal, _authKeytabPath);
+        UserGroupInformation tokenUser = realUser;
+        if (_authUser != null) {
+            log("Impersonate " + _authUser, null);
+            // attempt to impersonate token user, this verifies that the real-user is able to impersonate tokenUser
+            tokenUser = UserGroupInformation.createProxyUser(_authUser, tokenUser);
+        }
+        return tokenUser;
+    }
+    
+    private long getTokenRenewalInterval(UserGroupInformation tokenUser) throws IOException, InterruptedException {
+        Credentials creds = refreshTokens(tokenUser);
+        return tokenUser.doAs((PrivilegedExceptionAction<Long>) () ->
+            creds.getAllTokens()
+                    .stream()
+                    .map(token -> {
+                        try {
+                            long expiresAt = token.renew(PersistHdfs.CONF);
+                            long issuedAt = 0;
+                            TokenIdentifier ident = token.decodeIdentifier();
+                            if (ident instanceof AbstractDelegationTokenIdentifier) {
+                                issuedAt = ((AbstractDelegationTokenIdentifier) ident).getIssueDate();
+                            }
+                            return expiresAt - (issuedAt > 0 ? issuedAt : System.currentTimeMillis());
```

Comment:
What about a presence of one expired token where it's identifier is not `AbstractDelegationTokenIdentifier`. In that case, the method would return a negative number. Can it happen, is it a problem?

Reply:
The logic works this way: 
- we get a token and renew it (line 153) - which means we are only dealing with a valid token
- if the token isn't an instance of AbstractDelegationTokenIdentifier we will consider the current time as the time of renewal

the renewal intervals are long (hours) so we should never have a negative number here - however that case can be checked as well, will improve